kong/pdk/ctx.lua

```diff
@@ -87,16 +87,12 @@ local _CTX_CORE_KEY = {}
 --   kong.log(value) -- "hello world"
 -- end
 local function new(self)
-  local _CTX = {
-    -- those would be visible on the *.ctx namespace for now
-    -- TODO: hide them in a private table shared between this
-    -- module and the global.lua one
-    keys = setmetatable({}, { __mode = "v" }),
-  }
-
-
+  local _CTX = {}
   local _ctx_mt = {}
 
+  function _CTX.set(name, key)
```

Comment:
@thibaultcha, could you please check this PR as you probably understand the best the original design of this. Should it rather have `__newindex` or should it be solved otherway around, or approve the proposal?

Reply:
A note on item 2. above:

Since the key's key prefix should also be used when retrieving the key in the below `_ctx_mt.__index()` metamethod, we should localize it in this module, i.e.:
```lua
local fmt = string.format -- top of this module
local nctx_key_name = "kong_ctx_%s_key"


-- ...


local function new(self)
  local _CTX = {}
  local _ctx_mt = {}
  

  function _ctx_mt.set(name, key)
    ngx.ctx[fmt(nctx_key_name, name)] = key  
  end


  -- ... 


 -- same `fmt()` call in `__index()`
end
```

Reply:
And finally, a general note on the implications of this new design: while it does fix a blatant concurrency issue in the PDK's `kong.ctx` API, this new design makes the instantiation of multiple PDK instances share their `kong.ctx` state, which may or may not be the desired behavior. Since we do not instantiate multiple PDK instances (we have thought of doing so to introduce breaking changes yet preserve older plugins' backwards compatibility), this isn't an issue for now, just something to keep in mind.